onnxruntime/core/providers/cpu/fp16/fp16_conv.cc

```diff
@@ -0,0 +1,562 @@
+#include "core/providers/cpu/nn/conv_attributes.h"
+
+#include "contrib_ops/cpu/fused_activation.h"
+
+namespace onnxruntime {
+
+using ConvPadVector = ConvAttributes::ConvPadVector;
+
+/**
+ * @brief Convolution Operator for FP16 tensors
+ *
+ * With two optional fused operations:
+ *
+ * 1. Add
+ * It takes an extra (optional) input Sum, a tensor same shape as the output.
+ * Sum is added to the output tensor.
+ *
+ * 2. Activation
+ * It takes an operator attribute 'activation', which supplies the activation info.
+ *
+ * Add is performed BEFORE activation.
+ *
+ * The implementation runs faster with NHWC. By default, it converts NCHW to NHWC
+ * before processing, and convert the result back. It can take NHWC tensors directly.
+ * Use operator attribute 'channels_last' to specify that the data layout is NHWC.
+ *
+*/
+class FusedConvFp16 final : public OpKernel {
+ public:
+  FusedConvFp16(const OpKernelInfo& info) : OpKernel(info), conv_attrs_(info) {
+    ORT_ENFORCE(GetFusedActivationAttr(info, activation_).IsOK());
+    channels_last_ = (info.GetAttrOrDefault<int64_t>("channels_last", static_cast<int64_t>(0)) != 0);
+  }
+
+  Status Compute(OpKernelContext* context) const override;
+
+  Status PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+      /*out*/ bool& is_packed, /*out*/ PrePackedWeights* prepacked_weights) override;
+
+ private:
+
+  /**
+   * @brief Reorder filter data to facilitate compute.
+   *
+   *        Based on Conv operator spec, filters are organized as (M x C/group x kH x kW),
+   *        where C is the number of input channels, and kH and kW are the height and width
+   *        of the kernel, and M is the number of feature maps. We need to change it into
+   *        (kH x kW x C/group) x M, forming a matrix of M columns, where each kernel is a
+   *        single column in channel last format.
+   *
+   * @param input
+   * @param output
+   * @param output_channels  number of feature maps
+   * @param input_channels
+   * @param kernel_size      kH x kW
+  */
+  static void ReorderFilter(const MLFloat16* input,
+                            MLFloat16* output,
+                            size_t output_channels,
+                            size_t input_channels,
+                            size_t kernel_size) {
+    for (size_t k = 0; k < kernel_size; k++) {
+      for (size_t ic = 0; ic < input_channels; ic++) {
+        for (size_t oc = 0; oc < output_channels; oc++) {
+          size_t index = (oc * input_channels * kernel_size) + (ic * kernel_size) + k;
+          *output++ = input[index];
+        }
+      }
+    }
+  }
+
+  MLAS_ACTIVATION activation_;
+  ConvAttributes conv_attrs_;
+  bool channels_last_{false};
+  TensorShape W_shape_;
+  BufferUniquePtr packed_W_buffer_;
+  size_t packed_W_size_{0};
+  bool is_W_packed_{false};
+  BufferUniquePtr reordered_W_buffer_;
+};
+
+
+Status FusedConvFp16::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
```

Comment:
Can you please add a test like this - https://github.com/microsoft/onnxruntime/blob/32533dd1c290aa6025c584a705d3fa00fde1cd0a/onnxruntime/test/providers/cpu/math/matmul_test.cc#L297 to ensure that sharing of prepacked weights at a session level is kicking in ?

Reply:
added